### PR TITLE
Erasure code recovery should send additional reads if necessary

### PR DESCRIPTION
**Pull request by dzafman**
Fixes: http://tracker.ceph.com/issues/21382

### CODE REVIEW

**Comment by dzafman** [comment]
Test this again please

**Comment by dzafman** [comment]
Testing results
http://pulpito.ceph.com/dzafman-2017-09-22_10:23:26-rados-wip-21382-distro-basic-smithi/

2 unrelated failures out of 255 rados suite jobs:
1658401 saw valgrind issues
1658403 "2017-09-22 17:55:28.482691 mon.a mon.1 172.21.15.197:6789/0 101 : cluster [WRN] Health check failed: 1/3 mons down, quorum a,c (MON_DOWN)" in cluster log

**Comment by dzafman** [comment]
retest this please

**Comment by tchaikov** [comment]
sorry, i missed this PR. will review it early tomorrow.

File: qa/standalone/erasure-code/test-erasure-eio.sh [hunk 1]
```diff
@@ -281,10 +289,10 @@ function TEST_rados_get_subread_eio_shard_0() {
 
 function TEST_rados_get_subread_eio_shard_1() {
     local dir=$1
-    setup_osds || return 1
+    setup_osds 4 || return 1
```
Comment (tchaikov) [inline qa/standalone/erasure-code/test-erasure-eio.sh]: 3 OSDs would suffice for an erasure pool of "2+1" .

File: qa/standalone/erasure-code/test-erasure-eio.sh [hunk 2]
```diff
@@ -380,6 +380,35 @@ function TEST_rados_get_with_subreadall_eio_shard_1() {
...
+    local poolname=pool-jerasure
+    create_erasure_coded_pool $poolname 3 2 || return 1
+
+    rados_put $dir $poolname $objname || return 1
+    inject_eio ec data $poolname $objname $dir 0 || return 1
+
+    local -a initial_osds=($(get_osds $poolname $objname))
+    local last=$((${#initial_osds[@]} - 1))
+    # Kill OSD
+    kill_daemons $dir TERM osd.${initial_osds[$last]} >&2 < /dev/null || return 1
```
Comment (tchaikov) [inline qa/standalone/erasure-code/test-erasure-eio.sh]: nit, if we want to get the last element in a bash array, 
```sh
local last_osd=${initial_osds[-1]}
```
would suffice

File: qa/standalone/erasure-code/test-erasure-eio.sh [hunk 3]
```diff
@@ -380,6 +380,35 @@ function TEST_rados_get_with_subreadall_eio_shard_1() {
...
+    ceph osd down ${initial_osds[$last]} || return 1
+    ceph osd out ${initial_osds[$last]} || return 1
+
+    # Cluster should recovery this object
+
```
Comment (tchaikov) [inline qa/standalone/erasure-code/test-erasure-eio.sh]: nit, remove empty line. s/recovery/recover/

File: src/osd/ECBackend.cc [hunk 4]
```diff
@@ -1204,6 +1203,7 @@ void ECBackend::handle_sub_read_reply(
       set<int> want_to_read, dummy_minimum;
       get_want_to_read_shards(&want_to_read);
       int err;
+      // TODO: Should we include non-acting nodes here when for_recovery is set?
```
Comment (tchaikov) [inline src/osd/ECBackend.cc]: yeah, i think we should.

File: src/osd/ECBackend.cc [hunk 5]
```diff
@@ -1485,19 +1484,12 @@ void ECBackend::call_write_ordered(std::function<void(void)> &&cb) {
   }
 }
 
-int ECBackend::get_min_avail_to_read_shards(
+void ECBackend::get_shards_and_have(
```
Comment (tchaikov) [inline src/osd/ECBackend.cc]: might want to rename it to `get_all_avail_shards()`?

File: qa/standalone/erasure-code/test-erasure-eio.sh (hunk 1 above)
Comment (dzafman) [reply qa/standalone/erasure-code/test-erasure-eio.sh]: I'm taking 1 OSD down/out which can begin recovery to the 4th OSD.  Not sure this is critical to this test, but I'd rather be using a more realistic scenario.

Comment (tchaikov) [reply qa/standalone/erasure-code/test-erasure-eio.sh]: oh, i see.

**Pull request merged**